### ISSUE TITLE
Fire events

### ISSUE THREAD

**Issue by jonagoldman** [cybercog/laravel-eloquent-flag#33]
I see lots of ```// :TODO: Fire an event here``` in the source code. Is there a reason events are not fired?

**Comment by a-komarev** [comment]
Because I haven't created them because there were no requirement in my code yet. But it reminds me that it will be good to add events for the future.
You can make PR with events if you need of them.

**Comment by jonagoldman** [comment]
In order to make a PR a need some guidance.

1. Do you want one event per flag? (HasAcceptedAt/HasAcceptedAt --> AcceptedEvent, HasActiveFlag --> ActivatedEvent, ...)
2. Any naming convention for the event class? (Accepted, FlagAccepted, AcceptedEvent ...)
3. Use SerializesModels trait like in Laravel [default event](https://laravel.com/docs/5.5/events#defining-events)?

Something like this?

```php
<?php

namespace Cog\Flag\Events;

class Accepted
{

    public $model;

    /**
     * Create a new event instance.
     *
     * @param  Model $model
     * @return void
     */
    public function __construct(Model $model)
    {
        $this->model = $model;
    }
}
```

**Comment by a-komarev** [comment]
The main reason why I didn't added events from a start - because I don't know if its good to fire them if user don't want to. And the second reason I was thinking about possibility of customization\swapping event on user ones.

1. Yes, 1 type of flag = 1 event. `HasAcceptedAt` & `HasAcceptedFlag` should fire `ModelWasAccepted` event.
2. I think we can fire something like: `ModelWasAccepted`. It's pretty common event naming convention in Laravel community.
3. Thats the main question. Because not everybody need to Serialize model but a lot of people will like to have it. Thats why possibility to configure this stuff will be useful.

**Comment by jonagoldman** [comment]
Maybe an optional trait (besides the flag trait) and register the an observer in AppServiceProvider... (something like [this](https://laratrust.readthedocs.io/en/5.0/usage/events.html))?

```php
<?php

namespace App\Models;

use Cog\Flag\Traits\Classic\HasAcceptedFlag;
use Cog\Flag\Traits\HasEvents;

use Illuminate\Database\Eloquent\Model;

class Post extends Model
{
    use HasAcceptedFlag, HasEvents;
}
```

```php
<?php

namespace App\Providers;

use App\User;
use App\Observers\UserObserver;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{

    public function boot()
    {
        Post::observeFlags(PostObserver::class);
    }

}
```

**Comment by a-komarev** [comment]
Why do we need to add additional trait then? Because model can has many flags and then methods will raise a conflict?

**Comment by jonagoldman** [comment]
I will need to take a deeper look in the next few days. Maybe there is a simpler way, using Laravel's [```Model::observe()```](https://laravel.com/docs/5.5/eloquent#observers) method with custom events instead of a trait with the ```observeFlags()``` method.

**Comment by a-komarev** [comment]
@jonagoldman thank you and good luck! :}

**Comment by jonagoldman** [comment]
Found a solution using the same mechanism as the [Eloquent Model Events](https://laravel.com/docs/5.5/eloquent#events).
This way, we don't need to create lots of Event classes inside the package and we can leave it to the user to implement it.

Let's see for example how ```Cog\Flag\Traits\Classic\HasPublishedAtHelpers``` will look like after the modifications:

```php
namespace Cog\Flag\Traits\Classic;

use Carbon\Carbon;

trait HasPublishedAtHelpers
{

    // use model constructor to add 'published' and 'unpublished observable event names
    public function __construct(array $attributes = [])
    {
        parent::__construct($attributes);
        static::addObservableEvents(['published', 'unpublished']);
    }

    public function setPublishedFlag()
    {
        $this->published_at = Carbon::now();

        return $this;
    }

    public function unsetPublishedFlag()
    {
        $this->published_at = null;

        return $this;
    }

    public function isPublished()
    {
        return !is_null($this->published_at);
    }

    public function isUnpublished()
    {
        return !$this->isPublished();
    }

    public function publish()
    {
        $this->setPublishedFlag()->save();

        // fire the 'published' model event
        $this->fireModelEvent('published', false);
    }

    public function unpublish()
    {
        $this->unsetPublishedFlag()->save();

        // fire the 'unpublished' model event
        $this->fireModelEvent('unpublished', false);
    }
}
```

This is the only modification needed inside the package (3 lines of code for each type of flag).

Now, if the user wants to listen to the events, he will do it the same way it is done when listening to the default Eloquent model events (creating, updating, etc...).

For example, to listen to the 'published' model event:

In the Eloquent model that is implementing a flag trait:
```php
class MyModel extends Model
{

    use HasPublishedAt;

    /**
     * The event map for the model.
     *
     * @var array
     */
    protected $dispatchesEvents = [
        'published' => MyModelPublishedEvent::class,
    ];
}
```

He will need to create an observer:
```php
class MyModelObserver
{
    public function published(MyModel $myModel)
    {
        //         
    }
}
```

And then observe the events:
```php
class AppServiceProvider extends ServiceProvider
{
    public function boot()
    {
        MyModel::observe(MyModelObserver::class);
    }
}
```

Let me know what you think so I can make a PR.

**Comment by a-komarev** [comment]
@jonagoldman what a great finding! I think that's really good way to solve this task. The only one way I don't like in this implementation - `__constructor` overriding. What do you think about registering events on trait booting?

```php
/**
 * Boot the HasPublishedAtHelpers trait for a model.
 *
 * @return void
 */
public static function bootHasPublishedAtHelpers()
{
    $this->addObservableEvents(['published', 'unpublished']);
}
```

Will it work this way?

**Comment by a-komarev** [comment]
Ah... I see. It's only using this array if you want to customize events.

```php
/**
 * Fire a custom model event for the given event.
 *
 * @param  string  $event
 * @param  string  $method
 * @return mixed|null
 */
protected function fireCustomModelEvent($event, $method)
{
    if (! isset($this->dispatchesEvents[$event])) {
        return;
    }

    $result = static::$dispatcher->$method(new $this->dispatchesEvents[$event]($this));

    if (! is_null($result)) {
        return $result;
    }
}
```

So it wouldn't break default Eloquent events, because it will affect on them only if they are defined in this array.

```php
protected $dispatchesEvents = [
    'saving' => \App\Events\CustomSavingEvent::class,
];
```

**Comment by jonagoldman** [comment]
I don't like overriding the constructor either, the problem is that the trait booting method must be ```static``` and also ```addObservableEvents``` method is not static, so both options will not work:

```php
public static function bootHasPublishedAtHelpers()
{
    // this will throw FatalThrowableError: "Using $this when not in object context"
    $this->addObservableEvents(['published', 'unpublished']);

    // and this will throw Non-static method Illuminate\Database\Eloquent\Model::addObservableEvents() should not be called statically"
    static::addObservableEvents(['published', 'unpublished']);
}
```

The other option is make the user add the observables manually if he wants to use them:

```php
class MyModel extends Model
{

    use HasPublishedAt;

     /**
     * User exposed observable events.
     *
     * @var array
     */
    protected $observables = [
        'published',
    ];

    /**
     * The event map for the model.
     *
     * @var array
     */
    protected $dispatchesEvents = [
        'published' => MyModelPublishedEvent::class,
    ];

}
```

This way we don't need the constructor or the trait booting because we don't need the ```addObservableEvents``` method.
The user will need to manually set the ```$observables``` array, but already needs to set ```$dispatchesEvents``` anyway so I think this is not a problem.

**Comment by a-komarev** [comment]
Yes, i think that solution is better, because it's more explicit.

**Comment by jonagoldman** [comment]
Great. What about event names? I have come with a few but not sure about all of them.

| Trait name | Set Event | Unset Event |
| ---------- | ----- |  ----- |
| `HasAcceptedAt` | `accepted` | `unaccepted` |
| `HasAcceptedFlag` |  `accepted` | `unaccepted` |
| `HasActiveFlag` |  `activated` | `deactivated` |
| `HasApprovedAt` |  `approved` | `disapproved` |
| `HasApprovedFlag` |  `approved` | `disapproved` |
| `HasArchivedAt` |  `archived` | `unarchived` |
| `HasArchivedFlag` |  `archived` | `unarchived` |
| `HasClosedAt` |  `closed` | ? |
| `HasClosedFlag` | `closed` | ? |
| `HasDraftedAt` |  `drafted` | ? |
| `HasDraftedFlag` |  `drafted` | ? |
| `HasEndedAt` |  `ended` | ? |
| `HasEndededFlag` |  `ended` | ? |
| `HasExpiredAt` |  `expired` | ? |
| `HasExpiredFlag` | `expired` | ? |
| `HasInvitedAt` |  `invited` | `uninvited` |
| `HasInvitedFlag` |  `invited` | `uninvited` |
| `HasKeptFlag` |  ? | ? |
| `HasPublishedAt` |  `published` | `unpublished` |
| `HasPublishedFlag` | `published` | `unpublished` |
| `HasVerifiedAt` |  `verified` | `unverified` |
| `HasVerifiedFlag` | `verified` | `unverified` |

**Comment by a-komarev** [comment]
This question is really hard because of inconsistency of the antonyms for all of these words. Some of antonyms are even not exists. I've faced and described similar issue #11 before.

| HasClosedAt | `closed` | `opened` |
| HasClosedFlag | `closed` | `opened` |
| HasDraftedAt | `drafted` | `undrafted` |
| HasDraftedFlag | `drafted` | `undrafted` |
| HasEndedAt | `ended` | `unended` |
| HasEndededFlag | `ended` | `unended` |
| HasExpiredAt | `expired` | `unexpired` |
| HasExpiredFlag | `expired` | `unexpired` |
| `HasKeptFlag` | `kept` | `unkept` |

As you can see some flags naming is differ:
- `closed` opposite is `opened` and not `unclosed`
- `approved` opposite is `disapproved` and not `unapproved`
- `activated` opposite is `deactivated` and not `unactivated`

And I'm not sure is it better to change them to follow same naming convention, or leave it as it is.

**Comment by jonagoldman** [comment]
I have updated my https://github.com/cybercog/laravel-eloquent-flag/issues/33#issuecomment-374308278 to reflect changes.

I think the problem is that, for example, setting something as `closed` and then unsetting the status, is not the same as `closing` something and then `opening`.
I think `unclosed` and `unapproved` are real words and can be used, see: [unclosed](https://en.oxforddictionaries.com/definition/unclosed), [unapproved](https://en.oxforddictionaries.com/definition/unapproved), but for example `unactivated` is not a real word, and [undrafted](https://en.oxforddictionaries.com/definition/undrafted) definition is not to return something to a draft state...

**Comment by a-komarev** [comment]
You are right, Jona. If anything isn't `closed` it could mean that it's `opened` but in the same time it could mean that it's not `closed` and not `opened`, like state doesn't exist yet, or should use default value from application's model.
When I've started this package the task seemed to be easy and clear to understand. But then after more flags appeared I understood that it's not so easy. Because sometimes we need to have 2 flags in one model, for example `opened` and `closed`. Then we'll have an API naming issues. Because both of the flags will have same names: `open()` and `close()`. It should be solved in other way. I don't know exactly how at this moment.

I'm afraid if we'll dig into this issue, we are risking to never complete this feature. So this is very important thing, but it could be resolved in future versions. Since we are following SemVer, I'm OK with total refactoring and changing the behavior of the package later, because developers always could make a fork, or use an old version of it in any moment.

There is no word `unactivated` but there is `invactivated`. Historically this was a first flag of this package, and method naming was inspired from a lot of examples of `activated` flags from bunch of open source software.
Accordingly `undrafted` I've found somewhere over the internet discussion where some people said that `undraft` could mean "make draft complete". So I decided to use it, at least to keep naming convention consistent.

**Comment by jonagoldman** [comment]
So I guess is a matter of choosing the words that make sense now and if there is any problem it is always possible to change things later.
Let me know your final namings and I can make a PR.

**Comment by a-komarev** [comment]
I think current naming is good, at least it synchronized with methods naming.

**Comment by antonkomarev** [comment]
@jonagoldman sorry for delay. Released v4.0.0 with events firing!
